### PR TITLE
Switch to plain password auth for Dreamwidth

### PR DESCRIPTION
**Pull request by momijizukamori**
Changes related to this code push: https://dw-maintenance.dreamwidth.org/81112.html

### CODE REVIEW

**Comment by momijizukamori** [comment]
Github username is fine!

**Comment by momijizukamori** [comment]
Gah, one sec, Mark just pointed out I made some typos in the code, so I think the new DW route isn't actually getting hit.

File: js/background.js [hunk 1]
```diff
@@ -101,31 +101,46 @@ function loginas(this_account, callback)
 	var site_info = get_config_site_info(this_account.site_info);
 	var conn = new XMLHttpRequest();
 	console.log("Beginning the login dance...");
-	getLJchallenge(site_info.interfaceurl, function (challenge) {
-		var response = md5(challenge + this_account.password);
-		var params = "mode=sessiongenerate" +
+	// As of 4/26/20 Dreamwidth no longer supports challenge/response with passwords
+	if (site_info.naem == "Dreammwidth") {
```
Comment (alierak) [inline js/background.js:105]: naem?

Comment (momijizukamori) [reply js/background.js:105]: And 'Dreammwidth', because apparently I shouldn't write code before caffeine ;o;

**Comment by anathomical** [comment]
I'll fix this in the release branch I'm prepping to merge.  Testing with the fix shows it still seems to work just fine.

**Comment by momijizukamori** [comment]
Hmm, I'm getting errors trying to add accounts running with the typos fixed :|a Doing some more digging.

**Comment by momijizukamori** [comment]
Shit, all the passwords are stored as MD5 hashes :[ there's no upgrade route that doesn't involve users editing all their configs.

**Comment by anathomical** [comment]
Yeah, I was just running that test branch (the md5 thing).  I suspect that the API key route may be what we need to do.  Let me take a look at that to see what that looks like in terms of effort.  I know that invalidating accounts will result in a lot of user frustration.

**Comment by momijizukamori** [comment]
Agreed, I was hoping we could avoid that :\ I have a route for programmatically generating and returning an API key for a user that isn't in this code push but could maybe be squeezed in, but it requires an active login cookie.

**Comment by anathomical** [comment]
And since we won't be able to automatically obtain a valid active login cookie for accounts moving forward, we'll basically still be forcing users to re-enter their credentials, right?

**Comment by momijizukamori** [comment]
Yeah, it'd only work programmatically for accounts that have non-expired session cookies still in local storage :c

**Comment by kalinda99** [comment]
Unless it's different for others, none of those cookies will be valid because for some reason the DW cookies we use expire after 24 hrs (even though it says they're good for a year), we've been discussing a fix for this but hasn't implemented anything yet.

I suspect it's something DW does on their end since it doesn't do it for LJ or IJ

**Comment by momijizukamori** [comment]
I haven't dug into the session auth stuff, so I don't know.

Also, if either of you are discord users, most of DW development & support discussions take place there - https://discord.gg/TZwWUr

**Comment by anathomical** [comment]
So after doing some thought, the way forward may be to bite the bullet on BOTH downsides: refactoring for API keys AND accepting the fact that we're going to have to make users re-enter their passwords.

Implement a new account-creating flow for Dreamwidth that takes a a username and password, does a one-time `clear` auth to get a login cookie, then stores the API key (and whatever else is necessary for its use) in LocalStorage.

Now, this assumes that API keys don't have natural expiration/renewal cycles (because if they do, we'll probably want to allow for auto-renewal, and that would mean storing the username/password combo locally so that the same process could be repeated).

**Comment by anathomical** [comment]
@momijizukamori If you get a chance, can you share your code snippet that retrieves an API key?  I've made a quick short-term release that uses plaintext password login for now, but I'd like to replace that relatively soon with API keys since those are just going to be better security practice.

**Pull request merged**